### PR TITLE
Expand range of eps_step and eps in PGD attacks

### LINKED ISSUE #823: Expand range of eps_step (and potentially eps) in PGD
**Is your feature request related to a problem? Please describe.**
a) Enable the `eps_step` parameter to take values greater than `eps`. Currently, the max value is `eps`, though if taking a step from the boundary, the distance to the other side of the Lp ball is `2 * eps`. So, at the minimum, I would recommend allowing up to that value. However, since clipping and projection are performed after the step, I don't see a reason that `eps_step` couldn't be even larger.
b) Enable `eps_step = np.inf` (in particular, when `eps = np.inf`). If `eps` is infinite and `eps_step` is finite, it works fine, but if both are infinite, nans show up and cause errors. 
c) (Optional) Enable `eps = 0`. This would in principle be a no-op, but could be a useful test, e.g., checking for modifications to the input by the model.

**Describe the solution you'd like**
These would only require a couple of minor one-or-two line changes in `__init__` (a and c) and `_apply_perturbation` (b).

### PR DESCRIPTION
**Pull request by beat-buesser**
Signed-off-by: Beat Buesser <beat.buesser@ie.ibm.com>

# Description

This pull request adds support for `eps_step` larger than `eps` for all norms, `eps_step = np.inf` to enable stepping immediately towards the boundary of clip values in the direction of the gradient, and `eps=0` to zero any perturbation for debugging purposes.

Fixes #823

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
Hi @davidslater I think this should solve #823. What do you think, especially about the implementation of `eps_step = np.inf`?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 1]
```diff
@@ -311,15 +311,20 @@ def _apply_perturbation(
...
+        else:
+            eps_step = np.array(eps_step, dtype=ART_NUMPY_DTYPE)
+            x = x + torch.tensor(eps_step).to(self.estimator.device) * perturbation
+
+            if self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                x = torch.max(
+                    torch.min(x, torch.tensor(clip_max).to(self.estimator.device)),
+                    torch.tensor(clip_min).to(self.estimator.device),
+                )
```
Comment (davidslater) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:327]: Ditto on my comments above

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 2]
```diff
@@ -304,11 +304,16 @@ def _apply_perturbation(
...
             clip_min, clip_max = self.estimator.clip_values
-            x = tf.clip_by_value(x, clip_min, clip_max)
+            x = tf.where(perturbation < 0.0, clip_min, x)
+            x = tf.where(perturbation > 0.0, clip_max, x)
+        else:
+            x = x + tf.constant(eps_step, dtype=ART_NUMPY_DTYPE) * perturbation
+
+            if self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                x = tf.clip_by_value(x, clip_min, clip_max)
```
Comment (davidslater) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:316]: Ditto on my comments above

**Review by davidslater — CHANGES_REQUESTED** [review]
I think it generally looks good, but had some questions of how you're avoiding NaNs when eps_step is a vector.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 3]
```diff
@@ -185,20 +185,26 @@ def set_params(self, **kwargs) -> None:
...
             )
 
         if isinstance(self.eps, (int, float)):
-            if self.eps <= 0:
+            if self.eps < 0:
                 raise ValueError("The perturbation size `eps` has to be positive.")
         else:
-            if (self.eps <= 0).any():
+            if (self.eps < 0).any():
                 raise ValueError("The perturbation size `eps` has to be positive.")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:208]: Ok.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 4]
```diff
@@ -185,20 +185,26 @@ def set_params(self, **kwargs) -> None:
...
             raise TypeError(
-                "The perturbation size `eps` and the perturbation step-size `eps_step` must have the same type."
+                "The perturbation size `eps` and the perturbation step-size `eps_step` must have the same type of `int`"
+                ", `float`, or `np.ndarray`."
             )
 
         if isinstance(self.eps, (int, float)):
-            if self.eps <= 0:
+            if self.eps < 0:
                 raise ValueError("The perturbation size `eps` has to be positive.")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:205]: Ok.

File: art/attacks/evasion/fast_gradient.py [hunk 5]
```diff
@@ -281,24 +281,26 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
             raise TypeError(
-                "The perturbation size `eps` and the perturbation step-size `eps_step` must have the same type."
+                "The perturbation size `eps` and the perturbation step-size `eps_step` must have the same type of `int`"
+                ", `float`, or `np.ndarray`."
             )
 
         if isinstance(self.eps, (int, float)):
-            if self.eps <= 0:
+            if self.eps < 0:
                 raise ValueError("The perturbation size `eps` has to be positive.")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:301]: Ok.

File: art/attacks/evasion/fast_gradient.py [hunk 6]
```diff
@@ -281,24 +281,26 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
         else:
-            if (self.eps <= 0).any():
+            if (self.eps < 0).any():
                 raise ValueError("The perturbation size `eps` has to be positive.")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:304]: Ok.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py [hunk 7]
```diff
@@ -158,18 +158,6 @@ def _set_targets(self, x: np.ndarray, y: np.ndarray, classifier_mixin: bool = Tr
...
-
-        if (self.norm in ["inf", np.inf]) and (
-            (isinstance(self.eps, (int, float)) and self.eps_step > self.eps)
-            or (isinstance(self.eps, np.ndarray) and (self.eps_step > self.eps).any())
-        ):
-            raise ValueError("The iteration step `eps_step` has to be smaller than the total attack `eps`.")
-
-        if self.max_iter <= 0:
-            raise ValueError("The number of iterations `max_iter` has to be a positive integer.")
-
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py:161]: Yes, that makes sense.

File: art/attacks/evasion/fast_gradient.py [hunk 8]
```diff
@@ -377,11 +376,16 @@ def _apply_norm(grad, object_type=False):
...
+        if isinstance(eps_step, (int, float)) and eps_step == np.inf:
             clip_min, clip_max = self.estimator.clip_values
-            batch = np.clip(batch, clip_min, clip_max)
+            batch[perturbation < 0.0] = clip_min
+            batch[perturbation > 0.0] = clip_max
+        else:
+            batch = batch + eps_step * perturbation
+            if self.estimator.clip_values is not None:
+                clip_min, clip_max = self.estimator.clip_values
+                batch = np.clip(batch, clip_min, clip_max)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/fast_gradient.py:388]: No, actually I think this looks much better.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 1 above)
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:327]: Ok.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py]: Ok.

**Review by davidslater — APPROVED** [review]
LGTM
